File: AmazonPreparation/MergeKSortedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode resultList = null;
        ListNode resultListPointer = null;
        
        boolean allPointersNull = false;
        
        while(!allPointersNull) {
            int minNodeIndex = -1;
            for(int i=0; i<lists.length; i++){
                ListNode currentNode = lists[i];
                if(currentNode != null) {
                    if(minNodeIndex == -1 || currentNode.val < lists[minNodeIndex].val) {
                        minNodeIndex = i;
                    }
                }
            }
            
            if(minNodeIndex == -1) {
                allPointersNull=true;
            } else {
                if(resultList == null){
                    resultList = lists[minNodeIndex];
                    lists[minNodeIndex] = lists[minNodeIndex].next;
                    resultListPointer = resultList;
                } else {
                    resultListPointer.next = lists[minNodeIndex];
                    lists[minNodeIndex] = lists[minNodeIndex].next;
                    resultListPointer = resultListPointer.next;
                }
                resultListPointer.next = null;
            }
        }
        
        return resultList;
    }
}

File: AmazonPreparation/MergeKSortedLists.scala
import scala.collection.mutable.PriorityQueue

/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
    
    def mergeKLists(lists: Array[ListNode]): ListNode = {
        val priorityQueue: PriorityQueue[ListNode] = PriorityQueue()(
            Ordering.by((_: ListNode).x).reverse
        )
        
        for(l <- lists){
            var lt: ListNode = l
            while(lt != null){
                priorityQueue += lt
                lt = lt.next
            }
        }
        
        if(!priorityQueue.isEmpty){
            val root: ListNode = priorityQueue.dequeue()
            var node: ListNode = root

            while(!priorityQueue.isEmpty){
                node.next = priorityQueue.dequeue()
                node = node.next
                node.next = null
            }
            return root
        }
        
        null
    }
}